### ISSUE TITLE
Add built in noise functions to shaders

### ISSUE THREAD

**Issue by LightningAA** [godotengine/godot-proposals#2443]
### Describe the project you are working on
A game with shaders
### Describe the problem or limitation you are having in your project
I would like to create a simple **water shader** that has water that **displaces** a subdivided plane's **vertex y positions** so that it moves up and down. The problem is **Godot doesn't have any built in noise functions** for shaders and it makes it unnecessarily harder to do stuff like water shaders or grass shaders (where it looks like the grass sways in the wind). See for example [this](https://youtu.be/vm9Sdvhq6ho?t=409) water shader tutorial just animates the water with the sine function, which doesn't really look like water.

**Why not use an OpenSimplex noise texture?**
The problem with using a texture is it's not easy to animate over time. If 3d noise textures were added like suggested in #2261 then it would be possible to sample a 2d texture and use time as the third axis, but I can imagine that getting quite memory intensive and more importantly the texture would still repeat.
### Describe the feature / enhancement and how it helps to overcome the problem or limitation
See https://youtu.be/Vg0L9aCRWPE?t=610 for an example of how Unity's ShaderGraph does it. I think implementing simplex 1/2/3/4d noise would be enough.
### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
For my water shader, it could look something like:
```glsl
void vertex() {
    VERTEX.y += simplex_noise(vec3(UV, TIME));
}
```
### If this enhancement will not be used often, can it be worked around with a few lines of script?
Not with "a few lines", I found this library, https://github.com/curly-brace/Godot-3.0-Noise-Shaders but I think noise functions would be worth having built in. 
### Is there a reason why this should be core and not an add-on in the asset library?
I think this is a fairly common use case that many types of games could benefit from and I don't think it would be too much work to add

**Comment by LightningAA** [comment]
Perhaps a warning in the documentation for the function? I didn't have any issues that I can remember in Unity using noise to offset a couple thousand vertex's (and I have a several year old integrated graphics card).

**Comment by Xrayez** [comment]
Yes, I don't see a problem with performance if it's expected to perform slower as long as it's documented. It may be faster since it's going to be built-in. If users do use noise functions in shaders throughout Godot projects already, then we already have a performance problem. But baking noise as texture under the hood is nice to have.

**Comment by arkology** [comment]
> I really like the idea of adding these sorts of high-level functions to the visual shader as nodes. This is also super helpful for particle shaders.

There're already some plugins in asset library. For example, my plugin https://godotengine.org/asset-library/asset/589. I think, the problem is mostly in bad search support in asset library.

**Comment by LightningAA** [comment]
I did come up with a workaround for my specific water shader without using any 3rd party libraries: sample a seamless noise texture like
```glsl
uniform vec2 scroll_speed = vec2(0.05, 0.082238);

void vertex() {
    VERTEX.y += (texture(noise, UV + scroll_speed * TIME) * 2.0 - 1.0).r * height;
}
```


[2DNoiseWorkaround.zip](https://github.com/godotengine/godot-proposals/files/6118059/2DNoiseWorkaround.zip)

I would still rather have built in noise functions though.

**Comment by ChristianB84** [comment]
Unfortunately, the GLSL noise functions were usually not implemented and always returned 0.0. They have been deprecated since GLSL 4.4.

**Comment by cloewen8** [comment]
As a use case, I have a water shader that could use 3D noise. Alternatively, a tile-able "NoiseTexture3D" would work.

**Comment by Calinou** [comment]
> Alternatively, a tile-able "NoiseTexture3D" would work.

This is available since Godot 4.1: https://docs.godotengine.org/en/stable/tutorials/3d/volumetric_fog.html#using-3d-noise-density-textures
